text/3715-unsafe-derives-and-attrs.md

```diff
@@ -0,0 +1,126 @@
+- Feature Name: `unsafe_derives_and_attrs`
+- Start Date: 2024-10-22
+- RFC PR: [rust-lang/rfcs#3715](https://github.com/rust-lang/rfcs/pull/3715)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Allow declaring proc macro attributes and derive macros as unsafe, and
+requiring `unsafe` to invoke them.
+
+# Motivation
+[motivation]: #motivation
+
+Some traits place requirements on implementations that the Rust compiler cannot
+verify. Those traits can mark themselves as unsafe, requiring `unsafe impl`
+syntax to implement. However, trait `derive` macros cannot currently require
+`unsafe`. This RFC defines a syntax for declaring and using unsafe `derive`
+macros.
+
+This RFC also defines a syntax for declaring proc macro attributes as unsafe.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Derives
+
+When declaring a proc macro `derive`, you can add the `unsafe` parameter to the
+`proc_macro_derive` attribute to indicate that the derive requires `unsafe`:
+
+```rust
+#[proc_macro_derive(DangerousTrait, unsafe)]
+pub fn derive_helper_attr(_item: TokenStream) -> TokenStream {
+    TokenStream::new()
+}
+```
```

Comment:
Given that we're putting the `unsafe` around the derive macro name at the use site, what do we think about matching that at the def site?  E.g.:

```rust
#[proc_macro_derive(unsafe(DangerousDeriveMacro)]
pub fn derive_helper_attr(_item: TokenStream) -> TokenStream {
    TokenStream::new()
}
```

```suggestion
When declaring a proc macro `derive`, you can wrap the derive macro name with
`unsafe(..)` in the `proc_macro_derive` attribute to indicate that the derive
requires `unsafe`:

```rust
#[proc_macro_derive(unsafe(DangerousDeriveMacro)]
pub fn derive_helper_attr(_item: TokenStream) -> TokenStream {
    TokenStream::new()
}
```
```

Reply:
On one hand I like the symmetry, on the other hand (because we overloaded unsafe to have two meanings) I think this could be read as the derive macro itself discharging a safety obligation, whereas the first form makes it slightly clearer that this is just an unsafe derive.

Reply:
@carbotaniuman That's my feeling as well: `proc_macro_derive(unsafe(DangerousDeriveMacro))` reads to me as though the creation of it is unsafe, rather than it being a safe operation that creates an unsafe derive.

Reply:
I've added an alternative for this, along with rationale for why we shouldn't do it.